lookout/style/typos/metrics.py

```diff
@@ -0,0 +1,80 @@
+    """
+    Compose the set of k most probable correction candidates (tokens without probabilities).
+
+    :param corrections: List of corrections, sorted by probability.
+    :param k: Number of corrections to take.
+    :return: Set of k most probable correction tokens.
+    """
+    first_k = set()
+    for correction, _prob in corrections[:k]:
+        first_k.add(correction)
+    return first_k
+
+
+def get_scores(data: pandas.DataFrame, suggestions: Dict[int, List[Tuple[str, float]]],
+               mode: str = "correction", k: int = 1) -> Dict[str, float]:
+    """
+    Calculate the score of the solution of the specific typo correction problem.
+
+    Token is considered corrected, when the first suggestion doesn't match the token.
+    Supports three problems:
+    'detection': Typo is detected right: token is corrected when and only when it is not typo-ed.
+    'correction': Correctly spelled tokens should not be corrected. Typo-ed tokens should \
+                  contain the right correction among first k suggestions.
+    'on_corrected': Same as `correction`, but only the tokens, corrected by \
+                    the suggestions, are taken into account.
+    :param data: DataFrame which is indexed by Columns.Id and has columns Column.Token and \
+                 Column.CorrectToken.
+    :param suggestions: `{id : [(candidate, correct_prob)]}`, candidates are sorted \
+                        by correct_prob in a descending order .
+    :param mode: One of 'detection', 'correction', 'on_corrected'.
+    :param k: Number of the first suggested corrections to check. Used in modes \
+              'correction', 'on_corrected'.
+    :return: Scores of the suggestions.
+    """
+    y_true, y_pred = [], []
+    for i in data.index:
+        if mode == "on_corrected" and suggestions[i][0][0] == data.loc[i, Columns.Token]:
+            continue
+        typoed = data.loc[i, Columns.Token] != data.loc[i, Columns.CorrectToken]
+        if mode == "detection" or not typoed:
+            # If the word is not misspelled, model should not correct it in any mode
+            corrected_right = (suggestions[i][0][0] != data.loc[i, Columns.Token])
+        else:
+            corrected_right = (data.loc[i, Columns.CorrectToken] in first_k_set(suggestions[i], k))
+        y_pred.append(corrected_right)
+        y_true.append(data.loc[i, Columns.Token] != data.loc[i, Columns.CorrectToken])
+
+    precision, recall, f1, _ = precision_recall_fscore_support(y_true, y_pred, average="binary")
+    accuracy = accuracy_score(y_true, y_pred)
+    return {"accuracy": accuracy, "precision": precision, "recall": recall, "f1": f1}
+
+
+def print_all_scores(data: pandas.DataFrame, suggestions: Dict[int, List[Tuple[str, float]]],
+                     path: str = None) -> None:
+    """Print scores for suggestions in an easy readable way."""
+    file = None if not path else open(path, "w")
+    print("%-20s| %-10s| %-10s| %-10s| %-10s" %
```

Comment:
Can you please port the whole function to Jinja2. This may take a day or two, I know, but we are not in a hurry.

Reply:
In this PR or another one?

Reply:
As you prefer

Reply:
In this one, yes

Reply:
you can find examples here: https://github.com/src-d/style-analyzer/tree/master/lookout/style/format/templates

Reply:
@vmarkovtsev Should I focus on this, or do it in the meantime, while adding new code in next PRs?

Reply:
This is a "normal" priority.